pkg/crc/machine/cert-renewal.go

```diff
@@ -0,0 +1,230 @@
+
+	oc := oc.NewOcConfig(recoveryPod)
+	err = waitForRecoveryPod(oc)
+	if err != nil {
+		logging.Debugf("Error waiting for recovery apiserver to come up: %v", err)
+		return err
+	}
+
+	_, err = recoveryPod.runPodCommand("regenerate-certificates")
+	if err != nil {
+		logging.Debugf("Error running 'regenerate-certificates' command in the kube-apiserver-operator")
+		return err
+	}
+
+	forceRedeploymentPatch := fmt.Sprintf(`'{"spec": {"forceRedeploymentReason": "recovery-%s"}}'`, time.Now().Format(time.RFC3339))
+	for _, resourceType := range []string{"kubeapiserver", "kubecontrollermanager", "kubescheduler"} {
+		_, _, err := oc.RunOcCommand(
+			fmt.Sprintf("patch %s cluster", resourceType),
+			fmt.Sprintf("--patch=%s", forceRedeploymentPatch),
+			"--type=merge")
+		if err != nil {
+			logging.Warnf("Error forcing redeployment of %s: %v", resourceType, err)
+			return err
+		}
+	}
+
+	_, err = sshRunner.Run(fmt.Sprintf("sudo KUBECONFIG=%s /usr/local/bin/recover-kubeconfig.sh >kubeconfig && sudo mv kubeconfig /etc/kubernetes/kubeconfig", recoveryPod.GetKubeconfigPath()))
+	if err != nil {
+		logging.Debugf("Error invoking recover-kubeconfig.sh: %v", err)
+		return err
+	}
+
+	output, _, err := oc.RunOcCommand("get configmap kube-apiserver-to-kubelet-client-ca",
+		"-n openshift-kube-apiserver-operator",
+		"--template='{{ index .data \"ca-bundle.crt\" }}'")
+	if err != nil {
+		logging.Debugf("Error reading kubelet client CA: %v", err)
+		return err
+	}
+	/* Copy the ca-bundle data we just read to etc/kubernetes/kubelet-ca.crt */
+	_, err = sshRunner.Run(fmt.Sprintf("cat <<EOF | sudo tee /etc/kubernetes/kubelet-ca.crt\n%s\nEOF", output))
+	if err != nil {
+		logging.Debugf("Error writing kubelet client CA to /etc/kubernetes/kubelet-ca.crt: %v", err)
+		return err
+	}
+
+	_, _ = sshRunner.Run("sudo touch /run/machine-config-daemon-force")
+
+	_, err = sd.Stop("kubelet")
+	if err != nil {
+		return err
+	}
+	_, _ = sshRunner.Run("sudo rm -rf /var/lib/kubelet/pki /var/lib/kubelet/kubeconfig")
+	_, _ = sshRunner.Run("sudo crictl stopp $(sudo crictl pods -q)")
+	_, _ = sshRunner.Run("sudo crictl rmp $(sudo crictl pods -q)")
+
+	_, err = sd.Start("kubelet")
+	if err != nil {
+		logging.Debugf("Error restarting kubelet service: %v", err)
+		return err
+	}
+
+	/* 2 CSRs to approve, one right after kubelet restart, the other one a few dozen seconds after
+	approving the first one */
+	err = waitForPendingCsrs(oc)
+	if err != nil {
+		logging.Debugf("Error waiting for first pending CSR: %v", err)
+		return err
+	}
+	err = oc.ApproveNodeCSR()
+	if err != nil {
+		logging.Debugf("Error approving first pending CSR: %v", err)
+		return err
+	}
+
+	err = waitForPendingCsrs(oc)
+	if err != nil {
+		logging.Debugf("Error waiting for second pending CSR: %v", err)
+		return err
+	}
+	err = oc.ApproveNodeCSR()
+	if err != nil {
+		logging.Debugf("Error approving second pending CSR: %v", err)
+		return err
+	}
+
```

Comment:
Once this is done, we need to wait some more or the next steps won't work. It's not clear what we should be waiting on though nor how long :( this is why it's not in this iteration yet.